README.org
[[https://travis-ci.org/mssola/writer-mode][https://travis-ci.org/mssola/writer-mode.svg?branch=master]]

* writer-mode

THIS IS UNDER DEVELOPMENT AND EXPERIMENTAL

This is a simple and convenient minor mode that might be useful for
writers.

** Usage

*** TODO: describe workflow, ideas and shortcuts

** Configuration

*** TODO there are some variables and hooks which are not documented here. This will be done whenever I have something releaseable

- =writer-jump-to-first-headline= (default: =t=): it tells
  =writer-jump-to-outline= to jump into the closest first level headline. If you
  want to jump into the closest headline (even if it's deeper), then set this to
  =nil=. You can trust this default unless you are doing really weird things
  with headlines.
- =writer-forced= (default =nil=): this mode is disruptive, since it will
  attempt to delete the rest of the windows. By setting this to =nil=, it will
  ask the user whether this is really what they want before doing anything at
  all. If this is set to =t=, then it will always try to start this mode with
  the currently selected window and buffer without asking. It's recommended to
  set it to =t=, but since it's disruptive it doesn't take this value by
  default.
- =writer-pre-hook= (default =nil=): hook to be run before setting up the
  environment. Use this to add some special values before this mode does
  anything at all (e.g. =line-spacing=).
- =writer-notes-post-hook= (default =nil=): hook to be run after setting up the
  the right pane (notes section). Use this to further tune the right section of
  the workspace (e.g. creating a new window with =dired= mode in it).

** Dependencies

There are two optional dependencies:

- [[https://github.com/rnkn/olivetti][olivetti]]: when installed, it will be enabled automatically by this mode.
- [[https://github.com/joostkremers/writeroom-mode][writeroom-mode]]: when installed, it will be available by calling its related
  functions (e.g. =writer-room=).

** Contributing

Read the [[./CONTRIBUTING.org][CONTRIBUTING.org]] file.

** [[https://pbs.twimg.com/media/DJDYCcLXcAA_eIo?format=jpg&name=small][Changelog]]

Read the [[./CHANGELOG.org][CHANGELOG.org]] file.

** Appreciation

Thanks to [[https://github.com/frankjonen][@frankjonen]] for working on [[https://github.com/frankjonen/emacs-for-writers/][Emacs for Writers]]. This mode was developed
before he released this code, but out of frustration I abandoned this project
before it was finished. I retook the development for this project after he
submitted his code, after reconsidering some doubts that I was having. Moreover,
I also took some ideas from him (e.g. notes section).

Last but not least, thank you to all the GNU Emacs developers for providing such
a wonderful tool, and its ecosystem for providing useful packages such as
[[https://github.com/rnkn/olivetti][olivetti]] or [[https://github.com/joostkremers/writeroom-mode][writeroom-mode]].

** License

#+BEGIN_SRC text
Copyright (C) 2019 Miquel Sabaté Solà <mikisabate@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
#+END_SRC


CHANGELOG.org
#+STARTUP:showall

* Changelog

** 0.1 (under development)

- Users can make use of the =writer-jump= function in order to create an outline
  of the current document.
- There is a note section available. This can be disabled by setting the
  =writer-notes-hidden= customizable variable. Otherwise, this section can be
  toggled between shown/hidden by calling the =writer-notes-toggle= (or directly
  using the =writer-notes-hide= and =writer-notes-show= functions).
- This mode will be automatically enabled if the given org-mode document
  has =writer-mode:t= in the =#+PROPERTY= section.
- [[https://github.com/rnkn/olivetti][olivetti]] is an optional dependency, which is going to be enabled automatically
  by this mode if available.
- [[https://github.com/joostkremers/writeroom-mode][writeroom-mode]] is an optional dependency, which is going to be picked up by
  functions such as =writer-room= and =writer-room-quit= if available.


t/ci.sh
#!/bin/bash
# Copyright (C) 2019 Miquel Sabaté Solà <mikisabate@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# This is only here because the linters on GNU Emacs 25.x behave differently
# than in 26.x. I'm taking GNU Emacs 26.x as the source of truth for linters.

set -ex

if [ -z "$CI" ]; then
    exit 0
fi

cd "$( cd "$( dirname "$0" )/.." && pwd )"
if [ "$EVM_EMACS" = "emacs-25.1-travis" ] ||
       [ "$EVM_EMACS" = "emacs-25.2-travis" ] ||
       [ "$EVM_EMACS" = "emacs-25.3-travis" ]; then
    make git-validation
    make unit-test
else
    make test
fi


t/writer-tests.el
;;; writer-tests.el --- tests for writer-mode

;; Copyright (C) 2019 Miquel Sabaté Solà <mikisabate@gmail.com>
;;
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;; Author: Miquel Sabaté Solà <mikisabate@gmail.com>
;; Version: 0.1
;; Keywords: minor mode writer
;; URL: https://github.com/mssola/writer-mode

;;; Commentary:
;;
;; Main file of the test suite of this mode.
;; See README.org for documentation on how to run these tests.

;;; Code:

(require 'ert)

(require 'writer)

(provide 'writer-tests)

;;; writer-tests.el ends here


writer-org.el
;;; writer-org.el --- Further org-mode integration -*- lexical-binding: t; -*-
;;
;; Copyright (C) 2019 Miquel Sabaté Solà <mikisabate@gmail.com>
;;
;; Author: Miquel Sabaté Solà <mikisabate@gmail.com>
;; Version: 0.1
;; Package-Requires: ((emacs "25.1"))
;; Keywords: convenience files wp
;; URL: https://github.com/mssola/writer-mode
;;
;; This file is not part of GNU Emacs.
;;
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; This file provides additional integration capabilities with org-mode.
;; See README.org for documentation on how to use this minor mode.

;;; Code:

(require 'org)

;;; Utils

;; Function taken from https://emacs.stackexchange.com/a/21472.
(defun writer-org--global-props (&optional buffer keyword)
  "Get the plists of global org properties from the given BUFFER.
If not buffer was passed, then this function will call `current-buffer'.
Moreover, you can pass the KEYWORD to be used for the search.  It defaults to
'PROPERTY' if no keyword was specified."

  (with-current-buffer (or buffer (current-buffer))
    (org-element-map
        (org-element-parse-buffer)
        'keyword
      (lambda (el)
        (when (string-equal (org-element-property :key el) (or keyword "PROPERTY"))
          (nth 1 el))))))

(defun writer-org--prop-value (key)
  "Fetch the value of the property by the given KEY."

  (let ((key-re (writer-org--key-re key))
        (props (writer-org--global-props))
        ret
        val)
    (dolist (pr props)
      (save-match-data
        (and (string-match key-re (setq val (plist-get pr :value))))
        (setq ret (match-string 1 val))))
    ret))

(defun writer-org--key-re (key)
  "Fetch a regular expression for properties by the given KEY."

  (concat "^" (regexp-quote key) ":\\(\\w+\\)"))

;;; Enabling writer-mode

(defun writer-org-enable-p ()
  "Return non-nil if 'writer-mode' is set to 't' in an org property."

  (let ((enable (writer-org--prop-value "writer-mode")))
    (when (string= enable "t") t)))

(provide 'writer-org)

;;; writer-org.el ends here


writer-notes.el
;;; writer-notes.el --- Support for the notes section  -*- lexical-binding: t; -*-
;;
;; Copyright (C) 2019 Miquel Sabaté Solà <mikisabate@gmail.com>
;;
;; Author: Miquel Sabaté Solà <mikisabate@gmail.com>
;; Version: 0.1
;; Package-Requires: ((emacs "25.1"))
;; Keywords: convenience files wp
;; URL: https://github.com/mssola/writer-mode
;;
;; This file is not part of GNU Emacs.
;;
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; Code for handling the notes section.
;;
;;; Code:

(defcustom writer-notes-post-hook nil
  "Hook run just *after* setting up the notes section.
This hook is called without passing any argument.  Use this hook
to further tune the look and feel of the right section (e.g. opening up a new
window below with dired mode in it)."
  :type 'hook
  :group 'writer)

(defcustom writer-notes-hidden nil
  "Whether to display or not the notes section.
If this is set to t, then it will not open up the notes section when creating
the environment.  This variable is also used internally for functions that
hide/display this section.  This can always be reversed on runtime by calling
the 'writer-notes-show' function."
  :group 'writer
  :type '(choice (const :tag "Hide notes section" t)
                 (const :tag "Create the notes section" nil)))

(defconst writer-notes-file "notes.org"
  "The name of the file holding the notes.")

;;; Creation and initialization

(defun writer-notes-create-workspace (&optional mid-creation)
  "Create the environment from scratch.
For that, it assumes that there is only one window on the frame, and it then
allocates a 1/4 on the right for the notes section.  Otherwise, if MID-CREATION
was set to true, then this function assumes that two windows are already present
and that the cursor is currently at the rightmost window."

  (let ((dec (if mid-creation 1 0)))
    (unless writer-notes-hidden
      (dotimes (_ (- 3 dec))
        (split-window-right)
        (windmove-right))

      (writer-notes-init)
      (run-hook-with-args writer-notes-post-hook)
      (balance-windows)

      (dotimes (_ (- 2 dec))
        (windmove-left)
        (delete-window))
      (windmove-left))))

(defun writer-notes-init ()
  "Allocate the contents of the notes section."

  (find-file writer-notes-file t))

(defun writer-notes-init-and-balance ()
  "Initialize the notes space and balance the environment.
As for the environment, it assumes that there are 4 opened
vertical windows, the rightmost of which corresponds to the
current one."

  (unless writer-notes-hidden
    (writer-notes-init))
  (balance-windows)
  (windmove-left)
  (delete-window)
  (if writer-notes-hidden
      (delete-window)
    (windmove-left)))

;;; Show/hide section

(defun writer-notes-show ()
  "Show the notes section."

  (interactive)

  (if writer-notes-hidden
      (writer-notes--show)
    ;; If it's supposed to be shown but its window could not be located, then simply recreate the
    ;; environment by showing the notes section again. Otherwise, if there is a window containing
    ;; this buffer, just display an error message.
    (let ((cw (writer-notes--file-window)))
      (if cw
          (error "Notes section is already visible in window %s" cw)
        (writer-notes--show)))))

(defun writer-notes-hide ()
  "Hide the notes section if possible.
Do nothing if there is only one window available, and write an
error if we are supposed to have this section already hidden.
If none of these cases apply, it will hide the notes section."

  (interactive)

  (if writer-notes-hidden
      (error "Notes section should already be hidden")
    (progn
      ;; If there is only one window there are two options: either the current window is the notes
      ;; file, or it's another one. In the first case, then we simply need to kill the current
      ;; buffer. Otherwise we have to do nothing.
      (if (= (count-windows) 1)
          (let ((cw (writer-notes--file-window)))
            (setq writer-notes-hidden t)
            (if cw
                (kill-this-buffer)
              (error "There's only one window, doing nothing")))
        (writer-notes--hide)))))

(defun writer-notes-toggle ()
  "Toggle between showing/hiding the notes section."

  (interactive)

  (if writer-notes-hidden
      (writer-notes-show)
    (writer-notes-hide)))

(defun writer-notes--show ()
  "Recreate the notes section given the amount of windows visible.
This will error out if the amount of windows is more than 2, because this is a
scenario which is not supported right now by this mode."

  (setq writer-notes-hidden nil)
  (cond
   ((= (count-windows) 1) (writer-notes-create-workspace))
   ((= (count-windows) 2) (writer-notes--show-two))
   (t (progn
        (setq writer-notes-hidden t)
        (error "This is not possible with so many windows open")))))

(defun writer-notes--show-two ()
  "Recreate the notes section by assuming that there are two windows."

  (while (window-in-direction 'right)
    (windmove-right))
  (writer-notes-create-workspace t))

(defun writer-notes--file-window ()
  "Fetch the window containing the buffer with the notes file."

  (let (ret buf (name (expand-file-name writer-notes-file)))
    (dolist (window (window-list))
      (setq buf (window-buffer window))
      (if (string= (buffer-file-name buf) name)
          (setq ret window)))
    ret))

(defun writer-notes--kill-current-window (name)
  "Delete the current window and kill its buffer if it's the notes file.
Since this function is part of a loop and we don't want to compute the complete
file of the path over and over again, this function expects the caller to pass
the full path as the NAME."

  ;; Kill the buffer containing the notes file. Delete the window always.
  (when (string= name (buffer-file-name (current-buffer)))
    (kill-this-buffer))
  (delete-window)

  ;; It may happen that after deleting the window the cursor jumps to the left sibling. Move back to
  ;; the right.
  (if (window-in-direction 'right)
      (windmove-right)))

(defun writer-notes--kill-all-from-current-window ()
  "Delete all the vertical windows from the point of view of the current one.
This function will also kill the buffer containing the notes file."

  ;; This is implemented by moving up above, and then iterating while going down. On each iteration
  ;; it will call `writer-notes--kill-current-window'. Finally, it will move to the rightmost window
  ;; since this is the expected location of the main file.

  (while (window-in-direction 'above)
    (windmove-up))

  (let ((name (expand-file-name writer-notes-file)))
    (while (window-in-direction 'below)
      (writer-notes--kill-current-window name))
    ;; The iteration stopped when only one window was left, so we have to kill this one too.
    (writer-notes--kill-current-window name))

  (while (window-in-direction 'right)
    (windmove-right)))

(defun writer-notes--hide ()
  "Hide the notes section.
This is done by first locating the window containing the notes file.  It will
error out if this window could not be located."

  (let ((cw (writer-notes--file-window)))
    (if cw
        (progn
          (select-window cw t)
          (writer-notes--kill-all-from-current-window)
          (setq writer-notes-hidden t))
      (error "Could not locate window containing the file: %s" writer-notes-file))))

(provide 'writer-notes)

;;; writer-notes.el ends here


writer.el
;;; writer.el --- Main file for this mode -*- lexical-binding: t; -*-
;;
;; Copyright (C) 2019 Miquel Sabaté Solà <mikisabate@gmail.com>
;;
;; Author: Miquel Sabaté Solà <mikisabate@gmail.com>
;; Version: 0.1
;; Package-Requires: ((emacs "25.1"))
;; Keywords: convenience files wp
;; URL: https://github.com/mssola/writer-mode
;;
;; This file is not part of GNU Emacs.
;;
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; Main file of this mode.
;; See README.org for documentation on how to use this minor mode.

;;; Code:

(require 'org)
(require 'windmove)
(require 'writer-notes)
(require 'writer-org)

;;; Global & customizable variables.

(defvar writer--room-available nil
  "Whether writeroom is available.")

(defcustom writer-jump-to-first-headline t
  "Whether to jump to the first headline or to the closest one.
When you have multiple levels of heading and you press to jump inside of the
main window, sometimes you want to jump just to the main headline, or to the
closest one (even if it's a deeper level)."
  :group 'writer
  :type '(choice (const :tag "To the first headline" t)
                 (const :tag "To the closest headline" nil)))

(defcustom writer-forced nil
  "Force the execution of this mode.
If this is set to nil, then it will ask permission to the user to
start normally (i.e. delete the other windows).  This option is
given since users might be unaware that this mode is disruptive."
  :group 'writer
  :type '(choice (const :tag "Force execution" t)
                 (const :tag "Ask first" nil)))

(defcustom writer-pre-hook nil
  "Hook run just *before* setting up the environment.
This hook is called without passing any argument.  Use this hook
to add some special values before doing anything at all (e.g. a
different `line-spacing' than the one given here)"
  :type 'hook
  :group 'writer)

;;; Mode definition.

;;;###autoload
(define-minor-mode writer-mode
  "Toggle writer mode"

  :init-value nil
  :lighter " Writer"
  :global nil

  (if writer-mode
      (writer--enable)
    (writer--disable)))

;;; Functions for jumping between windows.

(defun writer-jump-to-outline ()
  "Switch to a cloned buffer's base buffer.
This function will also move the point to the cursor position in
the clone.  It has largely been taken from
https://emacs.stackexchange.com/a/9532, which has been written by Dan.
I've added the integration with `writer-jump-to-first-headline'."
  (interactive)

  ;; Quit 'writeroom-mode' if we are in it.
  (when (bound-and-true-p writeroom-mode)
    (writer-room-quit))

  (let ((buf (buffer-base-buffer)))
    (unless buf
      (error "You need to be in a cloned buffer!"))
    (let ((pos (point))
          (win (car (get-buffer-window-list buf))))

      ;; Select the right window and switch to the proper buffer.
      (if win
          (select-window win)
        (other-window 1)
        (switch-to-buffer buf))

      ;; Go to the position as given by the cloned buffer.
      (goto-char pos)

      ;; If we are to jump to a first level headline, hide everything from the
      ;; base and go to the beginning of the line.
      (when writer-jump-to-first-headline
        (progn
          (outline-hide-sublevels 1)
          (org-beginning-of-line)))

      ;; Finally, if the given point is not visible, show the needed context.
      (when (invisible-p (point))
        (outline-show-branches)))))

(defun writer-jump-from-outline ()
  "Open a window with a newly cloned buffer from the selected element of the outline."
  (interactive)

  (delete-other-windows)
  (split-window-right)
  (org-tree-to-indirect-buffer)
  (outline-hide-sublevels 1)

  (dotimes (_ 2)
    (windmove-right)
    (split-window-right))

  (windmove-right)
  (writer-notes-init-and-balance))

(defun writer-jump ()
  "Call either `writer-jump-from-outline' or `writer-jump-to-outline'."
  (interactive)

  (let ((buf (buffer-base-buffer)))
    (if buf
        (writer-jump-to-outline)
      (writer-jump-from-outline))))

;;; writeroom integration

(defun writer-room ()
  "Check some assumptions and enter 'writeroom-mode'."
  (interactive)

  (unless writer--room-available
    (error "You have to install writeroom-mode first"))
  (when (fboundp 'writeroom-mode)
    (writeroom-mode 1)))

(defun writer--room-force-quit (msg)
  "Quit 'writeroom-mode' and leave an error message as provided by `MSG'."
  (when (fboundp 'writeroom-mode)
    (writeroom-mode -1))
  (writer-error-and-disable msg))

(defun writer-room-quit ()
  "Gracefully quit 'writeroom-mode' and leave things as they were before entering."
  (interactive)

  (let ((buf (buffer-base-buffer))
        (cur (current-buffer)))

    ;; First of all, quit 'writeroom-mode'.
    (when (fboundp 'writeroom-mode)
      (writeroom-mode -1))

    ;; There are two cases here:
    ;;  1. If we were writing in a cloned buffer, then we have to switch to the
    ;;     base buffer (which is acting as the outline, so we have to hide the
    ;;     sublevels). Afterwards we have to remake the environment and finally
    ;;     switch to the cloned buffer in the middle.
    ;;  2. If this was not a cloned buffer, then we can simply restart the
    ;;     environment from scratch.
    (if buf
        (progn
          (switch-to-buffer buf)
          (outline-hide-sublevels 1)

          (dotimes (_ 3)
            (split-window-right)
            (windmove-right))

          (writer-notes-init-and-balance)
          (switch-to-buffer cur))
      (writer-notes-create-workspace))))

;; Enable and disable the mode, and related functions.

(defun writer-error-and-disable (msg)
  "Write the given error and disable this mode.
`MSG' is the message to be written in the Messages buffer."

  (writer-mode -1)
  (error msg))

(defun writer--enable ()
  "Set up writer mode for the current frame."

  (unless (derived-mode-p 'org-mode)
    (writer-error-and-disable "This minor mode only works in conjuction with Org mode"))

  (when (> (length (window-list)) 1)
    (unless writer-forced
      (unless (yes-or-no-p "This will delete the other windows. Are you sure? ")
        (writer-error-and-disable "User refused to start this mode"))))

  (if (require 'writeroom-mode nil t)
      (setq writer--room-available t)
    (message "writeroom-mode is not installed, disabling its integration"))
  (unless (require 'olivetti nil t)
    (message "olivetti is not installed, disabling its integration"))

  (writer-setup))

(defun writer-setup ()
  "Setup the environment.
This consists of two windows, were the one on the right takes a 1/4 of the total
available space.  It also sets values to variables such as `line-spacing' and it
starts olivetti mode with 100 columns."

  (setq line-spacing 3)
  (flyspell-mode 1)
  (setq global-hl-line-mode nil)
  ;; `set-fringe-mode' was introduced in GNU Emacs 25.2 and we should support GNU Emacs 25.1 too.
  (when (fboundp 'set-fringe-mode)
    (set-fringe-mode 0))

  (run-hook-with-args writer-pre-hook)
  (writer-notes-create-workspace)
  (when (fboundp 'olivetti-mode)
    (olivetti-mode 1))
  (when (fboundp 'olivetti-set-width)
    (olivetti-set-width 100)))


(defun writer-delete-other-windows-and-clones ()
  "Delete all the other windows and cloned buffers.
Note that this function assumes that we are on a window which contains the base buffer."

  ;; Iterate over the rest of the buffer list (we assume that the first one is
  ;; the one that we want to preserve), and remove all buffers that start with
  ;; the name of our current buffer.
  (let ((buffers (cdr (buffer-list)))
        (prefix (format "%s-" (buffer-name))))
    (while buffers
      (let ((cur (car buffers)))
        (when (with-current-buffer cur (string-prefix-p prefix (buffer-name cur)))
          (kill-buffer cur)))
      (setq buffers (cdr buffers))))

  ;; All the cloned buffers have been killed, now just delete the other windows.
  (delete-other-windows))

(defun writer-only-window-from-buffer (buf)
  "Go to the first window containing `BUF'and delete all the other windows.
If such a window doesn't exist, then the current one will be selected and the
given buffer will be placed on it."

  (let ((w (get-buffer-window buf (selected-frame))))
    (if w
        (select-window w)
      (switch-to-buffer buf)))
  (writer-delete-other-windows-and-clones))

(defun writer--disable ()
  "Remove writer mode for the current frame."

  (let ((buf (buffer-base-buffer)))
    (if buf
        (writer-only-window-from-buffer buf)
      (writer-delete-other-windows-and-clones)))

  ;; Disable for sure.
  (if writer-mode
      (writer-mode -1))

  ;; And finally kill this buffer. This is safe because all variables are local.
  (kill-this-buffer))

(defun writer--enable-maybe ()
  "Enable this mode depending on `writer-org-enable-p'.
That is, this function will enable this mode if the user set 'writer-mode:t' as
an 'org-mode' property."

  (when (and (writer-org-enable-p) (not writer-mode))
    (writer-mode)))

;;; Hooks

;; Automatically enable this mode if there is the org-mode property writer-mode is set to 't'.
(add-hook 'org-mode-hook 'writer--enable-maybe)

(provide 'writer)

;;; writer.el ends here
